back/engines/commercial/report_builder/app/services/report_builder/queries/visitors.rb

```diff
@@ -0,0 +1,145 @@
+      sessions = exclude_roles_if_needed(sessions, exclude_roles)
+
+      visitors_timeseries = sessions
+        .select(
+          "count(*) as visits, count(distinct(monthly_user_hash)) as visitors, date_trunc('#{resolution}', created_at) as date_group"
+        )
+        .group('date_group')
+        .order('date_group')
+        .map do |row|
+          {
+            visits: row.visits,
+            visitors: row.visitors,
+            date_group: row.date_group.to_date
+          }
+        end
+
+      stats = calculate_stats(sessions)
+
+      response = {
+        visitors_timeseries: visitors_timeseries,
+        visits_whole_period: stats[:visits],
+        visitors_whole_period: stats[:visitors],
+        avg_seconds_per_session_whole_period: stats[:avg_seconds_per_session],
+        avg_pages_visited_whole_period: stats[:avg_pages_visited]
+      }
+
+      # If compare_start_at and compare_end_at are present:
+      if compare_start_at.present? && compare_end_at.present?
+        compare_sessions = ImpactTracking::Session.where(created_at: compare_start_at..compare_end_at)
+        compare_sessions = apply_project_filter_if_needed(compare_sessions, project_id)
+        compare_sessions = exclude_roles_if_needed(compare_sessions, exclude_roles)
+
+        compare_stats = calculate_stats(compare_sessions)
+
+        response = {
+          **response,
+          visits_compared_period: compare_stats[:visits],
+          visitors_compared_period: compare_stats[:visitors],
+          avg_seconds_per_session_compared_period: compare_stats[:avg_seconds_per_session],
+          avg_pages_visited_compared_period: compare_stats[:avg_pages_visited]
+        }
+      end
+
+      response
+    end
+
+    def apply_project_filter_if_needed(sessions, project_id)
+      if project_id.present?
+        sessions_with_project = ImpactTracking::Pageview.where(project_id: project_id).select(:session_id)
+        sessions = sessions.where(id: sessions_with_project)
+      end
+
+      sessions
+    end
+
+    def exclude_roles_if_needed(sessions, exclude_roles)
+      if exclude_roles.present?
+        sessions = sessions
+          .where("highest_role IS NULL OR highest_role NOT IN (#{exclude_roles.map { |r| "'#{r}'" }.join(', ')})")
+      end
+
+      sessions
+    end
+
+    def calculate_stats(sessions)
+      # Total number of visits and visitors
+      visits = sessions.count
+      visitors = sessions.distinct.pluck(:monthly_user_hash).count
+
+      # Create new pageviews query for avg seconds per session and avg pages per session
+      pageviews = ImpactTracking::Pageview.where(session_id: sessions.select(:id))
+
+      # Calculate avg pages visited per session
+      # Or, if project filter is applied:
+      # Avg pages visited per session where someone visited the project during the session
+      avg_pages_visited = visits == 0 ? 0 : pageviews.count / visits.to_f
+
+      # Avg seconds per session
```

Comment:
This part is the most complicated. I am currently doing a lot of nested subqueries to arrive at the result, and the result is also an estimation (see comment below). Let me know if you have a more performant or just better way of calculating this!

Reply:
Maybe it would also be a good idea to cache this query? It should only be fired from the back office though, on the dashboards and in the report builder (not in the published reports)

Reply:
This is complex indeed. Can't think of a way to improve it. Not sure caching would help much given the input variables will be changing. I think go with it and then when it is deployed, find a platform that has the most page and session data and see how it performs.

In fact you could do that before you push live by connecting your local machine to a backup using Koen's sandbox cluster script.

Reply:
Think it should be OK.